shutdown_test.go

```diff
@@ -0,0 +1,120 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fx
+
+import (
+	"context"
+	"syscall"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestShutdown(t *testing.T) {
+	t.Run("ShutsDownApp", func(t *testing.T) {
+		var started, stopped bool
+		app := New(Invoke(func(l Lifecycle, s Shutdowner) {
+			l.Append(Hook{
+				OnStart: func(ctx context.Context) error {
+					started = true
+					return nil
+				},
+				OnStop: func(ctx context.Context) error {
+					stopped = true
+					return nil
+				},
+			})
+			l.Append(Hook{
+				OnStart: func(ctx context.Context) error { return s.Shutdown() },
+			})
+		}))
+
+		app.Run()
+		assert.True(t, started, "app wasn't started")
+		assert.True(t, stopped, "app wasn't stopped")
+	})
+
+	t.Run("BroadcastsToMultipleChannels", func(t *testing.T) {
+		var started, stopped bool
+		app := New(Invoke(func(l Lifecycle, s Shutdowner) {
+			l.Append(Hook{
+				OnStart: func(ctx context.Context) error {
+					started = true
+					return nil
+				},
+				OnStop: func(ctx context.Context) error {
+					stopped = true
+					return nil
+				},
+			})
+		}))
+
+		done1, done2 := app.Done(), app.Done()
+		startCtx, cancel := context.WithTimeout(context.Background(), app.StartTimeout())
+		defer cancel()
+		assert.NoError(t, app.Start(startCtx), "error in app start")
+
+		assert.NoError(t, app.shutdowner().Shutdown())
+
+		assert.Equal(t, syscall.SIGTERM, <-done1)
+		assert.Equal(t, syscall.SIGTERM, <-done2)
+
+		stopCtx, cancel := context.WithTimeout(context.Background(), app.StartTimeout())
+		defer cancel()
+		assert.NoError(t, app.Stop(stopCtx), "error in app stop")
+		assert.True(t, started, "app wasn't started")
+		assert.True(t, stopped, "app wasn't stopped")
+	})
+
+	t.Run("ErrorOnUnsentSignal", func(t *testing.T) {
+		var started, stopped bool
+		app := New(Invoke(func(l Lifecycle, s Shutdowner) {
+			l.Append(Hook{
+				OnStart: func(ctx context.Context) error {
+					started = true
+					return nil
+				},
+				OnStop: func(ctx context.Context) error {
+					stopped = true
+					return nil
+				},
+			})
+		}))
+
+		done := app.Done()
+		s := app.shutdowner().(*shutdowner)
+		s.broadcastSignal(syscall.SIGINT)
```

Comment:
Why do we need to use unexported API here? I think we should be able to use the `Populate` option to pull the `Shutdowner` out of the container without using the unexported `shutdowner()` method. We should also be able to fill the first channel by calling `Shutdown` and asserting that we didn't get an error back.

Reply:
True; I wanted to be able to broadcast a different signal — SIGINT rather than SIGTERM — just to extra-assert that the channel received the signal from something other than `.Shutdown()`. That said, I'm generally in agreement with not using unexported methods in tests, and I don't think we gain anything really valuable from sending a SIGINT